File: pyforms_web/web/static/pyformsjs/ControlMap.js
class ControlMap extends ControlBase {

    ////////////////////////////////////////////////////////////////////////////////
    init_control() {
        super.init_control();

        this.jquery_place().replaceWith(
            `<div id='${this.place_id()}' class='field control ControlMap' >
                ${this.properties.label_visible ? `<label for="${this.control_id()}">${this.properties.label}</label>` : ''}
                <div style="min-height: ${this.properties.min_height}px;" id='${this.control_id()}' placeholder='${this.properties.placeholder}' />
            </div>`
        );

        this.set_value(this.properties.value);

        if (this.properties.error)
            this.jquery_place().addClass('error');
        else
            this.jquery_place().removeClass('error');

        if (this.properties.required) this.set_required();



        // Delay the map initialization for rendering proposes.
        setTimeout(() => {
            this.initMap();
        }, 500);
    };

    ////////////////////////////////////////////////////////////////////////////////

    initMap() {
        const layers = this.properties.value;

        this.map = L.map(this.control_id()).setView(
            this.properties.center,
            this.properties.zoom
        );

        if (layers) {
            layers.forEach(l => {
                L.tileLayer(l.url, l.options).addTo(this.map);
            });
        }

        this.initEditPolygon();

        this.process_commands();
    }

    initEditPolygon() {
        this.editableLayers = new L.FeatureGroup();
        this.map.addLayer(this.editableLayers);

        var options = {
            position: 'topleft',
            draw: {
                polyline: this.properties.edit_polyline,
                polygon: this.properties.edit_polygon,
                marker: this.properties.edit_marker,
                circlemarker: this.properties.edit_circlemarker,
                circle: this.properties.edit_circle,
                rectangle: this.properties.edit_rectangle
            },
            edit: {
                featureGroup: this.editableLayers, //REQUIRED!!
                remove: true
            }
        };

        this.drawControl = new L.Control.Draw(options);
        this.map.addControl(this.drawControl);

        this.map.on(L.Draw.Event.CREATED, e => {
            this.editableLayers.addLayer(e.layer);
            this.basewidget.fire_event(this.name, 'update_control_event');
        });

        this.map.on(L.Draw.Event.EDITED, e => {
            this.basewidget.fire_event(this.name, 'update_control_event');
        });

        this.map.on(L.Draw.Event.DELETED, e => {
            this.basewidget.fire_event(this.name, 'update_control_event');
        });
    }

    process_commands() {
        if (this.properties.add_markers) {
            this.properties.add_markers.forEach((m, idx) => {
                const marker = L.marker(m.coordinate, m);
                this.editableLayers.addLayer(marker);
            });
        }

        if (this.properties.add_polygons) {
            this.properties.add_polygons.forEach(p => {
                const poly = L.polygon(p.coordinates, p);
                this.editableLayers.addLayer(poly);
            });
        }
        this.properties.add_markers = [];
        this.properties.add_polygons = [];

        if (this.properties.fitBounds && this.map && this.editableLayers.getBounds().isValid()) {
            this.map.fitBounds(this.editableLayers.getBounds());
        }
    }

    serialize() {
        const polygons = [];
        const markers = [];
        this.map.eachLayer(l => {
            if (l instanceof L.Polygon) {
                polygons.push(l.toGeoJSON())
            } else if (l instanceof L.Marker && l.options && l.options.pane === "markerPane") {
                markers.push(l.toGeoJSON())
            }

        });

        this.properties.polygons = polygons;
        this.properties.markers = markers;

        return super.serialize();
    }

    deserialize(data) {
        super.deserialize(data);
        if (this.properties.fitBounds && this.map && this.editableLayers.getBounds().isValid()) {
            this.map.fitBounds(this.editableLayers.getBounds());
        }
        if (this.map) {
            this.process_commands();
        }
    }

    /*
        enable() {
            this.drawControl.enable();
        }

        disable() {
            this.drawControl.disable();
        }
        */
}

File: pyforms_web/controls/control_map.py
import simplejson

from pyforms_web.controls.control_base import ControlBase


class ControlMap(ControlBase):

    def __init__(self, *args, **kwargs):
        """
        :param function on_enter_event: Event called when the Enter key is pressed.
        """
        super().__init__(*args, **kwargs)

        self.on_enter_event = kwargs.get('on_enter_event', self.on_enter_event)

        self._edit_polyline = kwargs.get('edit_polyline', False)
        self._edit_polygon = kwargs.get('edit_polygon', False)
        self._edit_marker = kwargs.get('edit_marker', False)
        self._edit_circlemarker = kwargs.get('edit_circlemarker', False)
        self._edit_circle = kwargs.get('edit_circle', False)
        self._edit_rectangle = kwargs.get('edit_rectangle', False)

        self._min_height = kwargs.get('min_height', 180);

        self.center = kwargs.get('center', [51.505, -0.09])
        self.zoom = kwargs.get('zoom', 10)
        self.fitBounds = kwargs.get('fitBounds', False)

        self._remove_layers = []
        self._add_layers = []
        self._add_markers = []
        self._add_polygons = []
        self.layers = {}
        self.markers = []
        self.polygons = []

    def init_form(self):
        return """new ControlMap('{0}', {1})""".format(
            self._name,
            simplejson.dumps(self.serialize())
        )

    def on_enter_event(self):
        """
        Event called when the Enter key is pressed
        """
        pass

    def clear_layers(self):
        self._remove_layers = [l for l in self.layers.values()]
        self.layers = {}
        self.mark_to_update_client()

    def add_layer(self, layer_url, options=None):

        if layer_url in self.layers:
            return

        layer = {'url': layer_url}
        if options:
            layer.update({'options': options})

        self._add_layers.append(layer)
        self.layers[layer_url] = layer
        self.mark_to_update_client()

    def remove_layer(self, layer):
        self._remove_layers.append(layer)
        self.layers.pop(layer['url'])
        self.mark_to_update_client()

    def add_mark(self, lat, long, **kwargs):
        marker = {'coordinate': [lat, long]}
        marker.update(**kwargs)
        self._add_markers.append(marker)
        self.markers.append(marker)

    def add_polygon(self, coordinates, **kwargs):
        polygon = {'coordinates': coordinates}
        polygon.update(**kwargs)
        self._add_polygons.append(polygon)
        self.polygons.append(polygon)

    def deserialize(self, properties):

        self.polygons = []
        for p in properties.get('polygons', []):
            poly = [(lat, long) for (long, lat) in p['geometry']['coordinates'][0]]
            self.polygons.append(poly)

        self.markers = []
        for m in properties.get('markers', []):
            coord =m['geometry']['coordinates']
            self.markers.append([coord[1], coord[0]])

        return super().deserialize(properties)

    def serialize(self):
        res = super().serialize()
        res.update({
            'center': self.center,
            'zoom': self.zoom,
            'fitBounds': self.fitBounds,
            'min_height': self._min_height,

            'add_markers': self._add_markers,
            'add_polygons': self._add_polygons,

            'add_layers': self._add_layers,
            'remove_layers': self._remove_layers,

            'edit_polyline': self._edit_polyline,
            'edit_polygon': self._edit_polygon,
            'edit_marker': self._edit_marker,
            'edit_circlemarker': self._edit_circlemarker,
            'edit_circle': self._edit_circle,
            'edit_rectangle': self._edit_rectangle,
        })
        self._add_markers = []
        self._add_polygons = []
        self._add_layers = []
        self._remove_layers = []

        self.fitBounds = False
        return res

